### ISSUE TITLE
cfreactor cannot be imported

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5088]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5088|
|Type|release blocker: regression|
|Created|2011-05-06 16:26:57Z|
|Branch|https://github.com/twisted/twisted/tree/cfreactor-import-5088|

```
>>> from twisted.internet import cfreactor
/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/zope/__init__.py:1: UserWarning: Module twisted was already imported from twisted/__init__.pyc, but /System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python is being added to sys.path
  __import__('pkg_resources').declare_namespace(__name__)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/exarkun/Projects/Twisted/trunk/twisted/internet/cfreactor.py", line 24, in <module>
    from twisted.internet.selectreactor import _NO_FILENO, _NO_FILEDESC
ImportError: cannot import name _NO_FILENO
```

<details><summary>Searchable metadata</summary>

```
trac-id__5088 5088
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__highest highest
milestone__Twisted_11_1 Twisted-11.1
branch__branches_cfreactor_import_5088 branches/cfreactor-import-5088
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1304699217000000 1304699217000000
changetime__1312370454000000 1312370454000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Looks like [#4539](https://github.com/twisted/twisted/issues/4539) is the culprit.  Why don't we have a failing buildbot?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Because `ImportError` gets turned into a skip automatically for the reactor builder tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

I guess I meant it as a rhetorical question.

Put more directly: one aspect of this fix should be a platform-specific "this should always run" thing which tells the OS X slaves to puke if they can't import this reactor for any reason.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Regression, so: highest priority.  I would just revert the other change but it looks like an important bugfix for its own right.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

And it needs to be completed before 11.1.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

I guess I'll take it over.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [31936]) Branching to 'cfreactor-import-5088'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

Anything else that needs to be done before this can go into review, other than adding a newsfile? This is only remaining blocker for 11.1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/cfreactor-import-5088).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @exarkun|
|-|-|

Thanks for turning the crank.  Land it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

I filed [#5216](https://github.com/twisted/twisted/issues/5216) to try to improve the reactor builder tests so we won't see something like this again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [32411]) Merge cfreactor-import-5088

Author: exarkun
Reviewer: glyph
Fixes: [#5088](https://github.com/twisted/twisted/issues/5088)

Fix a bad import in cfreactor which caused it to be unusable.  Also fix the
call-fileno-after-removal behavior which [#4539](https://github.com/twisted/twisted/issues/4539) fixed for the other reactors.
